docs/tutorial/frontend-tut.md

```diff
@@ -1,93 +1,158 @@
+In the [Calyx tutorial][calyx-tut] you wrote Calyx by hand.
+That is (probably) a good way to build character, but it's no way to live.
+In practice, you want a *frontend* that *compiles* to Calyx.
+
+This allows you to:
+1. Support domain-specific commands that Calyx does not.
+2. Automatically generate some of the kludge that Calyx requires.
+
+In this tutorial, we're going to learn all about this by building a compiler for a toy language.
+Meet MrXL.
+
+# MrXL Overview
+
+MrXL lets you define arrays and registers and then perform `map` and `reduce` operations.
+
+## A tiny example
+
+Here's a MrXL program in all its glory:
 
 ```
-{{#include ../../frontends/mrxl/test/dot.mrxl}}
+{{#include ../../frontends/mrxl/test/sos.mrxl}}
 ```
 
-We define the interface of program by specifying `input` and `output` arrays.
-Input arrays have their values populated by an external harness while the output arrays must be computed using the program.
+The program is short enough for us to pick apart line by line:
+1. We specify an array, `avec`, which will have four integers. The `input` keyword means that an external harness will populate the array.
+2. We specify `sos`, a register. The `output` keyword means that we will populate `sos` in our program.
+3. The `map` operation gets the values of `avec` and raises each to the second power. We stash the result in a new array, `squares`. The number `1` denotes a *parallelism factor* of 1, meaning that the operation is performed sequentially. We will improve this shortly.
+4. The `reduce` operation walks over `squares` and accumulates the result into a register. The parallelism factor is again `1`.
 
-A `map` expression iterates over multiple arrays of the same element and produces a new vector using the function provided in the body.
-In the above example, the `map` expression multiplies the values of `avec` and `bvec`.
-`map 1` states that the operation has a *parallelism factor* of 1 which means that the loop iterations are performed sequentially.
 
-`reduce` expressions walk over memories and accumulate a result into a register.
-In the above code snippet, we add together all the elements of `prodvec` and place them in a register named `dot`.
-Since the `reduce` parallelism factor is also 1, the reduction is performed sequentially.
+## Running our example
 
+Let's run this program.
 
-## Run a MrXL Program
+To begin, [install the MrXL command line tool][mrxldocs-install].
+(TK: I'd kinda like to inline those instructions here, so that there is less jumping around between files.
+I want folks to just install and get back here so we can interpret together, but there's a risk that folks will just begin to follow the other guide and go all the way to Verilog there.
+Also, the MrXL tutorial has a different running example.
+Thoughts on the inline?)
 
-Once we have [installed the mrxl command line tool][mrxldocs], we can run MrXL programs using [`fud`][fud].
+Now change directories to `calyx/frontends/mrxl` and run
```

Comment:
```suggestion
Now change directories to `calyx/frontends/mrxl` and run:
```
Just a suggestion, but I think it's nice to mark inline examples like this with colons.